### ISSUE TITLE
Asserting `call_count` of a `unittest.mock.Mock` object more than twice with at least one non-zero value incorrectly marks unreachable code

### ISSUE THREAD

**Issue by svioletg** [astral-sh/ty#4345]
### Summary

Title is a bit of a mouthful but I'm not sure how else to summarize it. I noticed an issue where ty seemed to incorrectly mark the rest of the code after `assert`ing the `call_count` attribute of a `unittest.mock.Mock` object twice, but only if at least one of those equality checks is against a number that isn't 0.

This, for example...

```python
mock = Mock()
assert mock.call_count == 0
assert mock.call_count == 0
assert mock.call_count == 0
assert mock.call_count == 0
assert mock.call_count == 0
# and so on...
```

...never gets marked as unreachable, which is expected. However, as soon as you change any of those to another number (negative or positive, as long as it's not 0), everything after that line is greyed out as unreachable. As a slightly more realistic minimal example, the script below will execute without issue as it should, but ty is still marking everything as unreachable after `assert mock.call_count == 1`, as shown in attached image.

```python
from unittest.mock import Mock


def test_mocked_calls() -> None:
    mock = Mock()

    print(mock.call_count)
    assert mock.call_count == 0
    mock(); print(mock.call_count)
    assert mock.call_count == 1
    mock(); print(mock.call_count)
    assert mock.call_count == 2
    mock(); print(mock.call_count)
    assert mock.call_count == 3

test_mocked_calls()
```

VSCode extension version is 2026.66.0.

Playground link: https://play.ty.dev/a5d3fa6f-1dd3-4a39-b749-7b0f2b679d1c

<img width="460" height="429" alt="Image" src="https://github.com/user-attachments/assets/1faf0f19-768d-473e-bc44-4bbd44f8f308" />

### Version

ty 0.0.69

**Comment by carljm** [comment]
Thanks for the report!

This is a consequence of a known limitation of attribute narrowing: a narrowed attribute type can be invalidated by a hidden change to the value of that attribute, which the type checker can't see. All Python type checkers do this kind of attribute narrowing even though it can be wrong, because the consequences of not doing it end up even worse in practice (you can assign an `int` to an `int | None` field and the type checker will still force you to consider that someone might have changed it back to `None` behind your back.)

That said, not all type checkers suffer from this problem in this case. Mypy does not, because it does not infer literal integer types from the assignments, just `int`, so it doesn't see the assertions as contradictory. I think a policy that we promote literal types inferred from attribute assignment narrowing would be defensible -- it wouldn't eliminate this problem entirely (e.g. it wouldn't fix the closely related issue in #2792), but it would make it show up much more rarely.

Alternatively, we could use a heuristic like pyright, which does suffer from this same problem in its type inference (it infers `Never` as the type of the `call_count` attribute) but doesn't emit an unreachable-code warning for it, because it explicitly excludes comparisons with attributes as a reason to consider a branch unreachable. It might be interesting to try a limited version of this, where we exclude comparison with a _narrowed_ attribute type as a cause for unreachability.

I kind of prefer the mypy approach, because I think literal inference is generally not critical, and the mypy heuristic fixes both type inference correctness and reachability, not just reachability.
